subprojects/snapshots/src/main/java/org/gradle/internal/snapshot/SearchUtil.java

```diff
@@ -37,23 +37,37 @@
      *         and only if the key is found.
      */
     public static <T> int binarySearch(List<T> sortedElements, SearchForComparator<T> searchForComparator) {
-        int low = 0;
-        int high = sortedElements.size() - 1;
+        int size = sortedElements.size();
+        switch (size) {
+            case 0:
+                return -1;
+            case 1:
+                T onlyElement = sortedElements.get(0);
+                int comparedToSearch = searchForComparator.compare(onlyElement);
+                return comparedToSearch == 0
+                    ? 0
+                    : comparedToSearch < 0
+                        ? -2
+                        : -1;
```

Comment:
This is confusing. Have you considered returning a cheap value object instead? Something like:

```groovy
class SearchResult {
    boolean found;
    int index;
}
```

I think the return value would be inlined by hotspot, so there's no performance impact, and the code would read a lot better.

Reply:
I just realized that this is from the contract of `Collections.binarySearch()`. I still think we should consider adding the type.

Reply:
I would like to keep the same semantics as `Collections.binarySearch()` for now.